driver-mapping/src/main/java/com/datastax/driver/mapping/ReflectionMapper.java

```diff
@@ -337,35 +225,19 @@ static DataType getSimpleType(Class<?> klass, Field f) {
                 }
 
-                if (field.getType().isAnnotationPresent(UDT.class)) {
+                if (TypeMappings.isMappedUDT(field.getType())) {
                     UDTMapper<?> udtMapper = mappingManager.getUDTMapper(field.getType());
                     return (ColumnMapper<T>) new UDTColumnMapper(field, position, pd, udtMapper);
                 }
 
                 if (field.getGenericType() instanceof ParameterizedType) {
-                    ParameterizedType pt = (ParameterizedType) field.getGenericType();
-                    Type raw = pt.getRawType();
-                    if (!(raw instanceof Class))
-                        throw new IllegalArgumentException(String.format("Cannot map class %s for field %s", field, field.getName()));
-
-                    Class<?> klass = (Class<?>)raw;
-                    Class<?> firstTypeParam = ReflectionUtils.getParam(pt, 0, field.getName());
-                    if (List.class.isAssignableFrom(klass) && firstTypeParam.isAnnotationPresent(UDT.class)) {
-                        UDTMapper<?> valueMapper = mappingManager.getUDTMapper(firstTypeParam);
-                        return (ColumnMapper<T>) new UDTListMapper(field, position, pd, valueMapper);
-                    }
-                    if (Set.class.isAssignableFrom(klass) && firstTypeParam.isAnnotationPresent(UDT.class)) {
-                        UDTMapper<?> valueMapper = mappingManager.getUDTMapper(firstTypeParam);
-                        return (ColumnMapper<T>) new UDTSetMapper(field, position, pd, valueMapper);
-                    }
-                    if (Map.class.isAssignableFrom(klass)) {
-                        Class<?> secondTypeParam = ReflectionUtils.getParam(pt, 1, field.getName());
-                        UDTMapper<?> keyMapper = firstTypeParam.isAnnotationPresent(UDT.class) ? mappingManager.getUDTMapper(firstTypeParam) : null;
-                        UDTMapper<?> valueMapper = secondTypeParam.isAnnotationPresent(UDT.class) ? mappingManager.getUDTMapper(secondTypeParam) : null;
-
-                        if (keyMapper != null || valueMapper != null) {
-                            return (ColumnMapper<T>) new UDTMapMapper(field, position, pd, keyMapper, valueMapper, firstTypeParam, secondTypeParam);
-                        }
+                    ExtractedType extractedType = ExtractedType.from(field, mappingManager);
+                    if (extractedType.containsMappedUDT) {
+                        // We need a specialized mapper to convert UDT instances in the hierarchy.
+                        return (ColumnMapper<T>)new NestedUDTMapper(field, position, pd, extractedType);
+                    } else {
+                        // The default codecs will know how to handle the extracted datatype.
+                        return new LiteralMapper<T>(field, extractedType.dataType, position, pd);
                     }
                 }
 
```

Comment:
Was thinking, couldn't this be:  

```java
return new LiteralMapper<T>(field, ExtractedType.from(field, mappingManager).dataType, position, pd);
```

This would remove the need for ```LiteralMapper#extractType(Field f)``` which has somewhat duplicate logic for parsing the type (which is also in ```ExtractedType``` constructor).

I couldn't find a way to exercise extractType's changes since there is a check here for ParameterizedType as well.